src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,355 @@
+          cardinality_sketch_factory=cardinality_sketch_factory)
+
+    return f
+
+  def __init__(self,
+               max_freq,
+               cardinality_sketch_factory,
+               random_seed,
+               underlying_set=None):
+    """Construct a Stratified sketch.
+
+    Args:
+      max_freq: the maximum targeting frequency level. For example, if it is set
+        to 3, then the sketches will include frequency=1, 2, 3+ (frequency >=
+        3).
+      random_seed: This arg exists in order to conform to
+        simulator.EstimatorConfig.sketch_factory.
+      cardinality_sketch_factory: A cardinality sketch factory.
+    """
+    SketchBase.__init__(self)
+    # A dictionary that contains multiple sketches, which include:
+    # (1) sketches with frequency equal to k, where k < max_freq;
+    # (2) a sketch with frequency greater than or equal to max_freq;
+    self.sketches = {}
+    self.seed = random_seed
+    self.max_freq = max_freq
+    self.cardinality_sketch_factory = cardinality_sketch_factory
+    self.underlying_set = underlying_set if underlying_set is not None else ExactMultiSet(
+    )
+
+  def create_sketches(self):
+    if(self.sketches != {}):
+      return
+    reversedict = {}
+    for k, v in self.underlying_set.ids().items():
+      reversedict.setdefault(min(v, self.max_freq), []).append(k)
+
+    for freq in range(1, self.max_freq):
+      self.sketches[freq] = self.cardinality_sketch_factory(self.seed)
+      if (freq in reversedict):
+        self.sketches[freq].add_ids(reversedict[freq])
+
+    # Initialize max_freq
+    max_key = str(self.max_freq) + '+'
+    self.sketches[max_key] = self.cardinality_sketch_factory(self.seed)
+    if (self.max_freq in reversedict):
+      self.sketches[max_key].add_ids(reversedict[self.max_freq])
+
+  def _destroy_sketches(self):
+    self.sketches = {}
+
+  def add(self, x):
+    if (self.sketches != {}):
+      self._destroy_sketches()
+      logging.warn(
+          """Tried to add ids after sketch creation, sketches are destroyed.""",
+          RuntimeWarning)
+    self.underlying_set.add(x)
+
+  @classmethod
+  def init_from_exact_multi_set(cls, max_freq, exact_multi_set,
+                                cardinality_sketch_factory, random_seed):
+    """Initialize a Stratified sketch from one ExactMultiSet.
+
+    Args:
+      exact_multi_set: ExactMultiSet object to use for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+    """
+    assert (cardinality_sketch_factory is
+            not None), ('cardinality_sketch is None')
+    stratified_sketch = cls(
+        max_freq=max_freq,
+        underlying_set=exact_multi_set,
+        cardinality_sketch_factory=cardinality_sketch_factory,
+        random_seed=random_seed)
+    stratified_sketch.create_sketches()
+
+    return stratified_sketch
+
+  @classmethod
+  def init_from_set_generator(cls, max_freq, set_generator,
```

Comment:
Do we expect this method to get called? It seems to me maybe not.

Reply:
Oh definitely, it is the main use case Matthew envisioned.